lib/grandpa/engine.go

```diff
@@ -0,0 +1,474 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/telemetry"
+)
+
+var errVotingRound = errors.New("voting round error")
+
+type finalizationHandler struct {
+	timeoutStop    time.Duration
+	grandpaService *Service
+	observableErrs chan error
+	stopCh         chan struct{}
+	handlerDone    chan struct{}
+}
+
+func newFinalizationHandler(service *Service) *finalizationHandler {
+	return &finalizationHandler{
+		timeoutStop:    5 * time.Second,
+		grandpaService: service,
+		observableErrs: make(chan error),
+		stopCh:         make(chan struct{}),
+		handlerDone:    make(chan struct{}),
+	}
+}
+
+func (fh *finalizationHandler) stopServices(finalizationEngine *finalizationEngine,
+	votingRound *handleVotingRound) error {
+	stopErr := make([]error, 2)
+	err := finalizationEngine.Stop()
+	if err != nil {
+		stopErr = append(stopErr, fmt.Errorf("stopping finalisation engine: %w", err))
+	}
+
+	err = votingRound.Stop()
+	if err != nil {
+		stopErr = append(stopErr, fmt.Errorf("stopping voting round: %w", err))
+	}
+
+	if len(stopErr) > 0 {
+		return stopErr[0]
+	}
+
+	return nil
+}
+
+var errStartingService = errors.New("starting service")
+
+func (fh *finalizationHandler) runFinalization() {
+	defer close(fh.handlerDone)
+
+	for {
+		err := fh.grandpaService.initiateRound()
+		if err != nil {
+			fh.observableErrs <- fmt.Errorf("%w: initiating round: %s", errStartingService, err)
+			return
+		}
+
+		finalizationEngine := newFinalizationEngine(fh.grandpaService)
+		votingRound := newHandleVotingRound(fh.grandpaService, finalizationEngine.actionCh)
+
+		finalizationEngErrsCh, err := finalizationEngine.Start()
+		if err != nil {
+			fh.observableErrs <- fmt.Errorf("%w: finalisation engine: %s", errStartingService, err)
+			return
+		}
+
+		handleVotingErrsCh, err := votingRound.Start()
+		if err != nil {
+			fh.observableErrs <- fmt.Errorf("%w: voting round: %s", errStartingService, err)
+			return
+
+		}
+
+		innerWg := sync.WaitGroup{}
+		innerWg.Add(2)
+
+		go func() {
+			defer innerWg.Done()
+			for err := range handleVotingErrsCh {
+				if err != nil {
+					fh.observableErrs <- fmt.Errorf("%w: %s", errVotingRound, err)
+					continue
+				}
+				return
+			}
+		}()
+
+		go func() {
+			defer innerWg.Done()
+			for err := range finalizationEngErrsCh {
+				if err != nil {
+					fh.observableErrs <- err
+					continue
+				}
+				return
+			}
+		}()
+
+		waitServicesCh := make(chan struct{})
+		go func() {
+			defer close(waitServicesCh)
+			innerWg.Wait()
+		}()
+
+		select {
+		case <-waitServicesCh:
+			err := fh.stopServices(finalizationEngine, votingRound)
+			if err != nil {
+				fh.observableErrs <- err
+				return
+			}
+
+		case <-fh.grandpaService.ctx.Done():
```

Comment:
this should be handled outside the `finalizationHandler`.  Once we receive the `Done` signal from the context, it should be calling `finalizationHandler.Stop()`.